keycloak_userprofile: new module

yeoldegrove opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

This adds a new keycloak module to manage user profiles.

See Keycloak documentation :
https://www.keycloak.org/docs/latest/upgrading/index.html#user-profile-changes
https://www.keycloak.org/docs/latest/server_admin/#_understanding-managed-and-unmanaged-attributes
https://www.keycloak.org/docs/latest/server_admin/#_user-profile-json-configuration

Fixes #8573 and #8579

<!--- Please do not forget to include a changelog fragment:
      https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-changelog-fragments
      No need to include one for docs-only or test-only PR, and for new plugin/module PRs.
      Read about more details in CONTRIBUTING.md.
      -->

##### ISSUE TYPE

- New Module/Plugin Pull Request


##### COMPONENT NAME
<!--- Write the SHORT NAME of the module, plugin, task or feature below. -->
keycloak_userprofile

##### ADDITIONAL INFORMATION

This is heavily based on `keycloak_realm_key` and `keycloak_user_federation`.

yeoldegrove commented:
Unit tests still do not work (hints welcome).
Some initial feedback would be appreciated, though.

felixfontein commented on tests/unit/plugins/modules/test_keycloak_userprofile.py:
```diff
@@ -0,0 +1,998 @@
+                with self.assertRaises(AnsibleExitJson) as exec_info:
+                    self.module.main()
+
+        self.assertEqual(len(mock_get_components.mock_calls), 2)
```
`ks.get_components` is only ever called once in the module (line 510, outside of a loop). Why do you expect it to be called twice?

yeoldegrove replied:
@felixfontein Thanks for the hint! After looking at it again I could make sense of the unit test ;)
Also I fixed all of the remaining sanity checks and reworked the integration tests.

I think this is good for a review.

yeoldegrove commented:
@felixfontein Anything else needed so this can be merged?

felixfontein commented on tests/sanity/ignore-2.18.txt:
```diff
@@ -2,6 +2,7 @@ plugins/modules/consul_session.py validate-modules:parameter-state-invalid-choic
 plugins/modules/homectl.py import-3.11  # Uses deprecated stdlib library 'crypt'
 plugins/modules/homectl.py import-3.12  # Uses deprecated stdlib library 'crypt'
 plugins/modules/iptables_state.py validate-modules:undocumented-parameter             # params _back and _timeout used by action plugin
+plugins/modules/keycloak_userprofile.py validate-modules:parameter-invalid
```
Why is this needed? Adding new entries here should be done with care, and this one looks like it is hiding a bug in the module.

yeoldegrove replied:
The concrete problem is the suboption `kc.user.profile.config` which is "so not python". I could of course abstract/rename this to be more pythonesque, but this is how it is named in keycloak's json representation.
Do you think it is valid to leave it or should it be changed?

felixfontein replied:
How about naming that option `kc_user_profile_config`, with an alias `kc.user.profile.config`? That way users can use both variants. The code will need to handle `kc_user_profile_config` though (and move its value to `kc.user.profile.config` when using `module.params['config']`), something like:
```
    config = module.params['config']
    if config:
        config = dict(config)  #  create a copy
        config['kc.user.profile.config'] = config.pop(`kc_user_profile_config`)
```
You're using `module.params['config']` (or its eqiuvalent `module.params.get('config')`) further down the code as well though, that needs to be adjusted to use this copy.

yeoldegrove replied:
Implemented your suggestion to name it `kc_user_profile_config` and now have aliases for a lot of (sub-)options.
Also added code that handles recursive camelCasel lists and dicts.

russoz replied:
I think it would be better not to have the name with the dots exposed to the Ansible user. It is not just the non- pythonesque aspect of it, but it would also create problems if that comes up in Jinja templating. Not to mention the need to create these exceptions in the sanity checking.
The problems outweigh the benefits, IMHO.

yeoldegrove replied:
Alright, removed the "doted" parameter option completely.

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,608 @@
+---
+module: keycloak_userprofile
+
+short_description: Allows managing Keycloak User Profiles.
```
This is the only description which shouldn't end with a period :)
```suggestion
short_description: Allows managing Keycloak User Profiles
```

yeoldegrove replied:
:+1:

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,608 @@
+                                'validations': dict(
+                                    type='dict',
+                                    # TODO: template this too but without passing null on unrequired empty key/value pairs
+                                    # options={
```
Basically if `options` isn't added now, it can never be added in the future.

What exactly is the problem here? That Ansible automatically adds `None` values for all fields the user did not provide? If that's the problem, you have to filter out these values yourself.

felixfontein commented:
Ping @yeoldegrove

yeoldegrove commented:
@felixfontein Finally had time to work on this again. I think I tackled everything you mentioned.

russoz commented on tests/unit/plugins/modules/test_keycloak_userprofile.py:
```diff
@@ -0,0 +1,1001 @@
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2021, Ansible Project
```
Copy paste quirks ;-)
```suggestion
# Copyright (c) 2024, Ansible Project
```

russoz reviewed (commented):
Hi @yeoldegrove , thanks for your contribution. Overall it does look good, there's just a couple of comments below, mostly minor issues.

yeoldegrove commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,668 @@
+        required: true
+    provider_id:
+        description:
+            - The name of the "provider ID" for the userprofile.
```
Adapted things to have more consistent naming.

yeoldegrove commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,668 @@
+        type: str
+    config:
+        description:
+            - The configuration of the User Profile Provider.
```
Adapted things to have more consistent naming.

yeoldegrove commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,668 @@
+def remove_null_values(data):
+    if isinstance(data, dict):
+        # Recursively remove null values from dictionaries
+        return {k: remove_null_values(v) for k, v in data.items() if v is not None}
```
:tada:

felixfontein replied:
Yep, we recently dropped Python 2.6 support :) (And also Python 2.7 support on the controller side - but for modules it's still supported.)

I'm looking forward to dropping Python 2.7 support for modules as well, IIRC that will happen somewhen next year (once the collection drops support for ansible-core 2.16).

yeoldegrove commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,668 @@
+    # Filter and map the parameters names that apply to the role
+    component_params = [x for x in module.params
+                        if x not in params_to_ignore and
+                        module.params.get(x) is not None]
```
:+1:

yeoldegrove commented on tests/sanity/ignore-2.13.txt:
```diff
@@ -5,6 +5,7 @@ plugins/lookup/etcd.py validate-modules:invalid-documentation
 plugins/lookup/etcd3.py validate-modules:invalid-documentation
 plugins/modules/consul_session.py validate-modules:parameter-state-invalid-choice
 plugins/modules/iptables_state.py validate-modules:undocumented-parameter             # params _back and _timeout used by action plugin
+plugins/modules/keycloak_userprofile.py validate-modules:parameter-invalid
```
Removed the ignores completely. This was initially about the doted `kc.user.profile.config` parameter.

yeoldegrove commented on tests/integration/targets/keycloak_userprofile/readme.adoc:
Set up the start of the container, too now.

yeoldegrove commented:
@russoz Thanks for the review! Reworked everything you mentioned and also re-shuffled a few things around with `partent_id` and `realm` and fixed some warnings and errors in the unit and integration tests.

felixfontein commented:
Looks ok to me. If nobody objects, I'll merge this for 9.4.0 on this weekend.

russoz commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,727 @@
+                            - ADMIN_VIEW
+
+notes:
+    - Currently only the provider_id of declarative-user-profile is tested.
+    - Currently only a single declarative-user-profile entry can exist. This entry can have multiple attributes, though.
```
These two lines should probably used semantic markup as well, but, the first one is quite redundant with the description in the option itself, on line 56. Maybe just move the second line to the description up there, put the V() around the value, and remove the first?

yeoldegrove replied:
@russoz rephrased the notes section and it uses semantic markup now as well.

felixfontein commented on changelogs/fragments/8651-keycloak-userprofile.yml:
Please remove this file, new modules should not come with changelog fragments. They are automatically announced via their `version_added` version.

yeoldegrove replied:
:zap:

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,727 @@
+
+    provider_id:
+        description:
+            - The name of the provider id for the key (supported value is V(declarative-user-profile)).
```
```suggestion
            - The name of the provider ID for the key (supported value is V(declarative-user-profile)).
```

yeoldegrove replied:
:+1:

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,727 @@
+                    unmanaged_attribute_policy:
+                        description:
+                            - Policy for unmanaged attributes.
+                            - Supported values are V(ENABLED), V(ADMIN_EDIT) and V(ADMIN_VIEW).
```
These are already listed under `choices`, so there's no need to list them again.
```suggestion
```

yeoldegrove replied:
:+1:

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,727 @@
+
+notes:
+    - Currently, only a single V(declarative-user-profile) entry is supported for O(provider_id) (design of the Keyckoak API).
+    - However, there can be multiple O(config.kc_user_profile_config[].attributes[]) entries.
```
I think this should be part of the previous note.
```suggestion
      However, there can be multiple O(config.kc_user_profile_config[].attributes[]) entries.
```

yeoldegrove replied:
:+1:

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,727 @@
+    state: absent
+    parent_id: master
+
+# Unmanaged attributes are user attributes not explicitly defined in the User Profile configuration. By default, unmanaged attributes are `Disabled`
+# and are not available from any context such as registration, account, and the administration console. By setting `Enabled`, unmanaged attributes are fully
+# recognized by the server and accessible through all contexts, useful if you are starting migrating an existing realm to the declarative User Profile
+# and you don't have yet all user attributes defined in the User Profile configuration.
```
Better limit the line length in EXAMPLES, otherwise the user has to scroll horizontally to be able to read this. Also better use regular quotes than backticks.
```suggestion
# Unmanaged attributes are user attributes not explicitly defined in the User Profile
# configuration. By default, unmanaged attributes are "Disabled" and are not
# available from any context such as registration, account, and the
# administration console. By setting "Enabled", unmanaged attributes are fully
# recognized by the server and accessible through all contexts, useful if you are
# starting migrating an existing realm to the declarative User Profile
# and you don't have yet all user attributes defined in the User Profile configuration.
```

yeoldegrove replied:
:+1:

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,727 @@
+      kc_user_profile_config:
+        - unmanagedAttributePolicy: ENABLED
+
+# By setting `Only administrators can write`, unmanaged attributes can be managed only through the administration console and API, useful if you have already
+# defined any custom attribute that can be managed by users but you are unsure about adding other attributes that should only be managed by administrators.
```
```suggestion
# By setting "Only administrators can write", unmanaged attributes can be managed
# only through the administration console and API, useful if you have already
# defined any custom attribute that can be managed by users but you are unsure
# about adding other attributes that should only be managed by administrators.
```

felixfontein commented on plugins/modules/keycloak_userprofile.py:
```diff
@@ -0,0 +1,727 @@
+      kc_user_profile_config:
+        - unmanagedAttributePolicy: ADMIN_EDIT
+
+# By setting `Only administrators can view`, unmanaged attributes are read-only and only available through the administration console and API.
```
```suggestion
# By setting "Only administrators can view", unmanaged attributes are read-only
# and only available through the administration console and API.
```

yeoldegrove replied:
:+1:

yeoldegrove commented:
@felixfontein Fixed your last suggestions.

felixfontein commented:
FYI @yeoldegrove you might be interested in https://forum.ansible.com/t/keycloak-modules-in-community-general/41746

russoz commented:
Apparently, upon reading this:

> Fixes #8573 and #8579 

GitHub associates only the first issue, the second one gets only a mention.
